### PR TITLE
Add the option to install the CUDA enabled nvidia docker runtime

### PR DESCRIPTION
**Pull request by chr15p**
Add a config option docker_runtime  that when set to "nvidia" will install the nvidia-docker2 container runtime and set it as default. This allows docker to run CUDA aware containers. 

This requires https://github.com/juju-solutions/layer-nvidia-cuda/pull/15

### CODE REVIEW

File: reactive/docker.py [hunk 1]
```diff
@@ -43,22 +44,36 @@
...
+dockerpackages = {'default':['docker.io'], 'upstream':['docker-engine'], 'nvidia':['docker-ce','nvidia-docker2','nvidia-container-runtime','nvidia-container-runtime-hook']}
+
+def holdall():
+    for k in dockerpackages.keys():
+        apt_hold(dockerpackages[k])
+
+def unholdall():
+    for k in dockerpackages.keys():
+        apt_unhold(dockerpackages[k])
+        #apt_unhold(k)
```
Comment (ktsakalozos) [inline reactive/docker.py]: Is that comments needed?

File: config.yaml [hunk 2]
```diff
@@ -15,7 +15,12 @@ options:
     type: boolean
     default: false
     description: |
-      Toggle installation from ubuntu archive vs the docker PPA
+      Toggle installation from ubuntu archive vs the docker PPA (DEPRECATED please use docker_runtime instead)
+  docker_runtime:
+    type: string
+    default: ""
```
Comment (ktsakalozos) [inline config.yaml]: I am not a vary big fan of this "whatever" maps to "deploy from apt". I would prefer if we had two more options "apt" -> "use docker from default apt repo" and "auto" -> "do what you think is best". We might get into trouble during upgrades if we leave this free text option.

File: reactive/docker.py [hunk 3]
```diff
@@ -74,67 +89,89 @@ def install():
...
     # Make with the adding of the users to the groups
     check_call(['usermod', '-aG', 'docker', 'ubuntu'])
 
 
-@when('config.changed.install_from_upstream', 'docker.ready')
+#@when('config.changed.install_from_upstream', 'docker.ready')
+@when('config.changed.docker_runtime', 'docker.ready')
 def toggle_docker_daemon_source():
-    ''' A disruptive toggleable action which will swap out the installed docker
+    ''' A disruptive
```
Comment (hyperbolic2346) [inline reactive/docker.py]: Maybe clean this comment up a little? A disruptive reaction to config changing that will ...

File: reactive/docker.py [hunk 4]
```diff
@@ -187,6 +224,46 @@ def install_from_upstream_apt():
...
+
+    ## mkdir -p /etc/apt/sources.list.d
+    if not os.path.isdir('/etc/apt/sources.list.d'):
+        os.makedirs('/etc/apt/sources.list.d')
+    ## Write the docker source file to the apt sources.list.d directory.
+    with(open('/etc/apt/sources.list.d/docker.list', 'w+')) as stream:
+        stream.write("\n".join(deb))
+
+    apt_update(fatal=True)
+    ## apt-get install -y docker-ce nvidia-docker2
```
Comment (hyperbolic2346) [inline reactive/docker.py]: remove

File: reactive/docker.py [hunk 5]
```diff
@@ -74,67 +89,89 @@ def install():
...
+            hookenv.log("Removing package(s): %s"%",".join(dockerpackages[k]))
+            apt_unhold(dockerpackages[k])
+            apt_purge(dockerpackages[k])
+            remove_state('docker.ready')
+            remove_state('docker.available')
     else:
         hookenv.log('Not touching packages.')
 
 
+
```
Comment (hyperbolic2346) [inline reactive/docker.py]: I think flake8 will complain about this extra space.

File: reactive/docker.py [hunk 6]
```diff
@@ -187,6 +224,46 @@ def install_from_upstream_apt():
     apt_install(['docker-engine'], fatal=True)
 
 
+def install_from_nvidia_apt():
+    ''' Install cuda docker from the nvidia apt repository. This is a pyton adaptation of
+    the shell script found at https://get.docker.com/ '''
+    status_set('maintenance', 'Installing docker-engine from Nvidia PPA.')
+
+    ## Get the server and key in the apt-key management tool.
```
Comment (ktsakalozos) [inline reactive/docker.py]: Could we try to reuse code either in the apt-layer or in the charm helpers for this repo management? The logstash charm has a config variable for the repo https://github.com/juju-solutions/layer-logstash/blob/85d85715b38f81b2e72dea30b5463aaf2284c5ef/config.yaml  and uses charmhelpers.fetch package to load the repo https://github.com/juju-solutions/layer-logstash/blob/85d85715b38f81b2e72dea30b5463aaf2284c5ef/reactive/logstash.py#L27 https://pythonhosted.org/charmhelpers/api/charmhelpers.fetch.html#charmhelpers.fetch.configure_sources . There is also the apt layers add_source https://git.launchpad.net/layer-apt/tree/lib/charms/apt.py#n37 would it make sense to have the right repos for each arch we support. The drawback is that there will be a lot of config parameters. The good thing is that we would have these configurable so we will have a proper way to address issues like: https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/452 . What do others think about this?

File: reactive/docker.py [hunk 7]
```diff
@@ -74,67 +89,89 @@ def install():
...
+    host.service_restart('docker')
     hookenv.log('Docker installed, setting "docker.ready" state.')
     set_state('docker.ready')
 
     # Make with the adding of the users to the groups
     check_call(['usermod', '-aG', 'docker', 'ubuntu'])
 
 
-@when('config.changed.install_from_upstream', 'docker.ready')
+#@when('config.changed.install_from_upstream', 'docker.ready')
```
Comment (ktsakalozos) [inline reactive/docker.py]: Changing the install_from_upstream config will have no effect.

**Comment by ktsakalozos** [comment]
These came up in some offline chats. There is some concern regarding the user experience we are offering. 

The first question is what we want the default behaviour to be in the presence of GPU capable hardware?
If we want to enable GPU support by default we would have to install nvidia drivers (nvidia layer), switch to the nvidia-docker daemon (docker layer), apply proper labels to the nodes at the kubernetes-worker. Reasons not to enable GPU by default may be, we do not trust* the nvidia-docker daemon; we do not want workers downloading the nvidia stack (some GBs) by default.

Also, there seems to be a strong relation between the docker and the nvidia layer. It does not make sense to use the nvidia-docker if the nvidia drivers are not installed. Some of the configs in the nvidia layer and this patch seem to be redundant (https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/config.yaml). It seems reasonable the docker layer to build on top of the nvidia layer (included in layer.yaml). Doing so might mean we need to fork the docker layer since the current docker layer is used by other charms as well. For example, the docker registry charm (that uses a docker registry container) would never need to deploy nvidia drivers and use the nvidia-docker.

Last thing. Enabling GPU support through a config variable would do so across all worker nodes. An action would allow us to enable GPU on certain workers. But this PR is not working towards this path. We could consider having such an action on the worker layer.

@wwwtyro , @Cynerva , @kwmonroe your thoughts?


(*) Haven't checked, but could there be compatibility and/or stability issues with the nvidia-docker?

**Comment by hyperbolic2346** [comment]
I am +1 to have a configuration variable to control this that defaults to yes/auto. The variable says to install all the nvidia stuff iff a GPU is detected. Basically a "I want to run GPU workloads, do whatever is necessary to make that happen". I think we should retain the magic for the user and not force them to know what enabling GPUs in kubernetes requires.

**Comment by Cynerva** [comment]
> The first question is what we want the default behaviour to be in the presence of GPU capable hardware?

I would advocate for a consistent, predictable user experience. Installing the nvidia stack and using an entirely different implementation of Docker are fairly invasive changes. We don't want to surprise users with that. Let them enable it explicitly.

> Also, there seems to be a strong relation between the docker and the nvidia layer. It does not make sense to use the nvidia-docker if the nvidia drivers are not installed. Some of the configs in the nvidia layer and this patch seem to be redundant

I wouldn't really call them redundant since they address two different components: drivers, and docker. If we try to hide the details and combine all this into a `enable_gpu_support=true` config or whatever, it could bite us later on when somebody wants to mix and match components in a way we didn't expect.

My experience with CDK has been that hiding details like this initially makes everything seem nice and easy (yay!) but eventually _something_ comes along to shatter our initial expectations, and we're forced to expose the details anyway. See api-extra-args and friends.

> Last thing. Enabling GPU support through a config variable would do so across all worker nodes. An action would allow us to enable GPU on certain workers. But this PR is not working towards this path. We could consider having such an action on the worker layer.

The problem with making this an action is that it breaks the scaling functionality in Juju, e.g. `juju add-unit`. I don't think that's worth it.

IMO, if the user wants to mix GPU and non-GPU nodes, then they should have two different applications in Juju with different configuration and constraints.

**Comment by Cynerva** [comment]
> I think we should retain the magic for the user and not force them to know what enabling GPUs in kubernetes requires.

I'm not opposed to this kind of magic as long as the lower level details remain accessible to those who need them. I haven't seen a good way to do this in charms. Maybe conjure-up?

**Comment by chr15p** [comment]
The only issue I see with this is that config options are set globally but
interpreted locally.

So "apt" means everything runs docker.io, "nvidia" means everything runs
nvidia-docker, and auto means each box makes its own decision. So the
"auto" setting will set up a mixed cluster and there is no other way of
doing that.

Thats not necessarily a bad thing, its just an oddity.

chris

**Comment by ktsakalozos** [comment]
@Cynerva, if I understand you correctly, you are +1 to having the docker layer build on the nvidia layer. You would also want to have the option to a) skip nvidia driver installation and b) set the nvidia driver version. I am mentioning these two configs because the current nvidia layer have the same two configs for cuda. Note that currently the nvidia driver version is fixed to 375 (https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/reactive/cuda.sh#L101) and the config options exposed by the nvidia layer (https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/config.yaml) have to do with cuda which is (probably) not needed anymore.

Parallel discussion on the nvidia layer PR: https://github.com/juju-solutions/layer-nvidia-cuda/pull/15

**Comment by Cynerva** [comment]
> @Cynerva, if I understand you correctly, you are +1 to having the docker layer build on the nvidia layer.

Yeah, I don't see a problem with that.

> You would also want to have the option to a) skip nvidia driver installation

Yes. This part is important to me because it gives operators control over what the charm does or does not do. They can disable it and manually install their own drivers, for example - not super clean, but it's an option they otherwise would not have at all.

>  and b) set the nvidia driver version.

This one's more of a nice-to-have, given the above.

> Note that currently the nvidia driver version is fixed to 375 (https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/reactive/cuda.sh#L101) and the config options exposed by the nvidia layer (https://github.com/juju-solutions/layer-nvidia-cuda/blob/master/config.yaml) have to do with cuda which is (probably) not needed anymore.

Oh, derp, I didn't realize the driver and cuda were different things. Thanks for explaining.

**Comment by chr15p** [comment]
On 14 March 2018 at 08:25, Konstantinos Tsakalozos <notifications@github.com
> wrote:

> Note that currently the nvidia driver version is fixed to 375 (
> https://github.com/juju-solutions/layer-nvidia-cuda/
> blob/master/reactive/cuda.sh#L101)
>
Erm. Yes but. Installing nvidia-375 actually gets you nvidia_384.ko,  and
then we install cuda without a version lock which pulls in nvidia-390
(which appears to be the latest) as a dependency, removing nvidia-384

A consequence of this in that sometimes (always?) removing 384 doesn't
unload the active kmod leaving the cuda libraries and the kmod out of sync,
until a reboot or rmmod. This causes cuda dependent containers to fail. I
try to work around this by reloading the nvidia kmod in lines 298-301 of my
patched reactive/docker.py

Parallel discussion on the nvidia layer PR: juju-solutions/layer-nvidia-

**Comment by ktsakalozos** [comment]
As agreed we merge this to get ready for the poc and we will return to the issues discussed here.

**Pull request merged**